### PR TITLE
 Interface with flowR's Server

### PR DESCRIPTION
**Pull request by Ellpeck**
For now, this is also just a cleaned up copy of the original extension

### CODE REVIEW

**Comment by Ellpeck** [comment]
(See #31 for plans on cleaning this up massively)

File: src/extension.ts [hunk 1]
```diff
@@ -32,6 +33,19 @@ export function activate(context: vscode.ExtensionContext) {
 			void vscode.window.showTextDocument(doc)
 		}
 	}))
+
+	context.subscriptions.push(vscode.commands.registerCommand('vscode-flowr.session.connect', () => {
+		flowrSession.destroy()
```
Comment (EagleoutIce) [inline src/extension.ts:38]: why not an "already connected" warning?

File: src/extension.ts [hunk 2]
```diff
@@ -32,6 +33,19 @@ export function activate(context: vscode.ExtensionContext) {
...
+		flowrSession = new FlowrServerSession(outputChannel, diagnostics)
+	}))
+	context.subscriptions.push(vscode.commands.registerCommand('vscode-flowr.session.disconnect', () => {
+		if(flowrSession instanceof FlowrServerSession) {
+			flowrSession.destroy()
+			flowrSession = new FlowrInternalSession(outputChannel, diagnostics)
+		}
+	}))
+
+	context.subscriptions.push(new vscode.Disposable(()=> flowrSession.destroy()))
```
Comment (EagleoutIce) [inline src/extension.ts:48]:
```suggestion
	context.subscriptions.push(new vscode.Disposable(() => flowrSession.destroy()))
```

Why does this even pass linting :P

File: src/flowr/internal-session.ts [hunk 3]
```diff
@@ -93,15 +90,21 @@ export class FlowrInternalSession {
...
 				tags:     [vscode.DiagnosticTag.Unnecessary]
 			})
 		}
-		this.diagnostics.set(uri, diagnostics)
-		this.outputChannel.appendLine('slice: ' + JSON.stringify([...result.slice.result]))
-		return result.reconstruct.code
+		return ret
+	}
+
+	public static fixEncoding(text: string) {
```
Comment (EagleoutIce) [inline src/flowr/internal-session.ts:103]: this is definitely hacky and - i think - no longer necessary after we dropped xmlparsedata!

File: src/flowr/internal-session.ts [hunk 4]
```diff
@@ -84,7 +69,19 @@ export class FlowrInternalSession {
 			return a.location.start.line - b.location.start.line || a.location.start.column - b.location.start.column
 		})
 
-		const diagnostics: vscode.Diagnostic[] = []
+		this.outputChannel.appendLine('slice: ' + JSON.stringify([...result.slice.result]))
```
Comment (EagleoutIce) [inline src/flowr/internal-session.ts:72]: we should only do diagonstics like this when desired - we may want to add a configuration which allows this.

File: src/flowr/internal-session.ts [hunk 5]
```diff
@@ -84,7 +69,19 @@ export class FlowrInternalSession {
...
+		this.diagnostics.set(uri, FlowrInternalSession.createDiagnostics(document, range, pos, sliceElements))
+		return result.reconstruct.code
+	}
+
+	public static getPositionAt(position: vscode.Position, document: vscode.TextDocument): vscode.Range | undefined {
+		const re = /([a-zA-Z0-9._:])+/
```
Comment (EagleoutIce) [inline src/flowr/internal-session.ts:78]: again, a really.... intersting way to get "the beginning" of a token - which does, for example, not work if the token is escaped. so we have two options

1) make this much more sophisticated and invest like 20 hours
2) handle the issue on the flowr side allowing a slicing point to be specified "looseley" by position (i.e. add another slicing point format which does not require the first position

i am for 2

File: src/flowr/internal-session.ts [hunk 6]
```diff
@@ -24,12 +21,10 @@ export class FlowrInternalSession {
 		void this.shell.usedRVersion().then(version => {
 			this.outputChannel.appendLine(`Using R shell: ${JSON.stringify(version)}`)
 		})
-		process.on('exit', () => {
```
Comment (EagleoutIce) [inline src/flowr/internal-session.ts:27]: In a way we should still have them, but link to destroy. we want a clean exit (even though, probably, the node runtime would cover us here. we want this to stay explicit).

Comment (EagleoutIce) [reply src/flowr/internal-session.ts:27]: But, theoretically, we can move this to the extension

File: src/flowr/server-session.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+import type { FlowrMessage } from '@eagleoutice/flowr/cli/repl'
+import type { FileAnalysisResponseMessageJson } from '@eagleoutice/flowr/cli/repl/server/messages/analysis'
+import type { SliceResponseMessage } from '@eagleoutice/flowr/cli/repl/server/messages/slice'
+import type { NodeId} from '@eagleoutice/flowr'
+import { visitAst } from '@eagleoutice/flowr'
+import type { SourceRange } from '@eagleoutice/flowr/util/range'
+import { isNotUndefined } from '@eagleoutice/flowr/util/assert'
+import { FlowrInternalSession } from './internal-session'
+
+export class FlowrServerSession {
```
Comment (EagleoutIce) [inline src/flowr/server-session.ts:12]: proposal: we should offer an abstract FlowrClient class as part of flowr which provides the basic logic like `connect(<address>, <port>)` which handles hello `request slicing` etc.

File: src/flowr/server-session.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+			if(n.location) {
+				idToLocation.set(n.info.id, n.location)
+			}
+		})
+
+		const sliceResponse = await this.sendCommandWithResponse<SliceResponseMessage>({
+			'type':      'request-slice',
+			'id':        String(this.idCounter++),
+			'filetoken': '@tmp',
+			'criterion': [`${pos.line + 1}:${pos.character + 1}`]
```
Comment (EagleoutIce) [inline src/flowr/server-session.ts:98]: again a mapping which we might want to outsource into a function as this just reads weird imho.

File: src/flowr/internal-session.ts (hunk 6 above)
Comment (Ellpeck) [reply src/flowr/internal-session.ts:27]: (also do this on `SIGINT` is what we want)

File: src/flowr/internal-session.ts (hunk 5 above)
Comment (Ellpeck) [reply src/flowr/internal-session.ts:78]: why are you criticizing your own code as if it was my fault that it's bad <3

File: src/extension.ts (hunk 1 above)
Comment (Ellpeck) [reply src/extension.ts:38]: added to #31

File: src/flowr/internal-session.ts (hunk 3 above)
Comment (Ellpeck) [reply src/flowr/internal-session.ts:103]: #32

File: src/flowr/internal-session.ts (hunk 5 above)
Comment (Ellpeck) [reply src/flowr/internal-session.ts:78]: https://github.com/Code-Inspect/flowr/issues/704

File: src/flowr/server-session.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+		this.socket.on('error', e => this.outputChannel.appendLine(`flowR server error: ${e.message}`))
+		this.socket.on('data', str => this.handleResponse(String(str)))
+	}
+
+	public destroy(): void {
+		this.socket.destroy()
+	}
+
+	private currentMessageBuffer = ''
+	handleResponse(message: string): void {
```
Comment (EagleoutIce) [inline src/flowr/server-session.ts:34]: we should definitely switch to the websocket version of flowr :sparkles:

**Review by EagleoutIce — APPROVED** [review]

**Pull request merged**